iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -17,45 +17,149 @@
 
 #include "iceoryx_posh/internal/roudi/service_registry.hpp"
 
+#include <algorithm>
+
 namespace iox
 {
 namespace roudi
 {
-void ServiceRegistry::add(const capro::IdString_t& service, const capro::IdString_t& instance)
+cxx::expected<ServiceRegistry::Error> ServiceRegistry::add(const capro::ServiceDescription& serviceDescription) noexcept
 {
-    cxx::set::add(m_serviceMap[service].instanceSet, instance);
+    if (!serviceDescription.isValid())
+    {
+        return cxx::error<Error>(Error::SERVICE_DESCRIPTION_INVALID);
+    }
+
+    // Forbid duplicate service descriptions entries
+    for (auto& element : m_serviceDescriptionVector)
+    {
+        if (element == serviceDescription)
+        {
+            return cxx::error<Error>(Error::SERVICE_DESCRIPTION_ALREADY_ADDED);
+        }
+    }
+
+    if (!m_serviceDescriptionVector.push_back(serviceDescription))
+    {
+        return cxx::error<Error>(Error::SERVICE_REGISTRY_FULL);
+    }
+    m_serviceMap.insert({serviceDescription.getServiceIDString(), m_serviceDescriptionVector.size() - 1});
+    m_instanceMap.insert({serviceDescription.getInstanceIDString(), m_serviceDescriptionVector.size() - 1});
+    return cxx::success<>();
 }
 
-void ServiceRegistry::remove(const capro::IdString_t& service, const capro::IdString_t& instance)
+bool ServiceRegistry::remove(const capro::ServiceDescription& serviceDescription) noexcept
 {
-    cxx::set::remove(m_serviceMap[service].instanceSet, instance);
+    if (!serviceDescription.isValid())
+    {
+        return false;
+    }
+
+    bool removedElement{false};
+    uint64_t index = 0U;
+    for (auto iterator = m_serviceDescriptionVector.begin(); iterator != m_serviceDescriptionVector.end();)
+    {
+        if (m_serviceDescriptionVector[index] == serviceDescription)
+        {
+            m_serviceDescriptionVector.erase(iterator);
+            removedElement = true;
+            // There can be not more than one element
+            break;
+        }
```

Comment:
Let's assume we have multiple publisher and only one of them does a `STOP_OFFER` then the service gets removed from the service registry, even if there still are publishing publishers?

Reply:
We may need multiset like tracking, but this should not be done in the prefixtree nor a map. We can have a count in a struct that we store in a container as the m_serviceDescriptionVector now (this is the wrong container for the job in the long run anyway, we should use the `ObjectPool`). The removing would decrease the associated count and only if it drops to 0 it is actually removed.

Storing actual copies multiple times would surely be the wrong approach though.
A vector is bad here, all the following elements will be moved upon removal. This will also completely lead to wrong results with the indexing, since it invalidates the indices stored in the maps/tries later.

This is a major problem in the current algorithm. Deleting anything from `m_serviceDescriptionVector` should not change the positions of other elements currently in the vector. It only should be logically removed (i.e. invalidated), not erased. The `ObjectPool` is made to deal with something like this. It works, but could be optimized and the code made a little nicer (in another PR).

Reply:
Right, we need a container with a ref-counting like `ServiceDescription`. Could be an array of `cxx::optional<{ServiceDescripton, Counter}>` and a FIFO with the free indices. Similar to what you do in the PrefixTree. As intermediate step we could have an array of `cxx::optional<ServiceDesctiption>` and one with `int16_t` for the counter. Finding a free slot would be quite fast since the array is small and packed on adjacent cache lines ... we could also use the object pool, though.

Reply:
Good catch @elBoberido! I added a plain and simple solution using `cxx::pair` holding the `ServiceDescription` and a `ReferenceCounter`. We can beautify it once we have decided on the final data structure.